src/core/reader/types/mod.rs

```diff
@@ -201,6 +201,85 @@ impl WasmReadable for FuncType {
     }
 }
 
+/// <https://webassembly.github.io/spec/core/binary/instructions.html#binary-blocktype>
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum BlockType {
+    Empty,
+    Returns(ValType),
+    Type(u32),
+}
+
+impl WasmReadable for BlockType {
+    fn read(wasm: &mut WasmReader) -> Result<Self> {
+        if wasm.peek_u8()? as i8 == 0x40 {
+            // Empty block type
+            let _ = wasm.read_u8().unwrap_validated();
+            Ok(BlockType::Empty)
+        } else if let Ok(val_ty) = ValType::read(wasm) {
+            // No parameters and given valtype as the result
+            Ok(BlockType::Returns(val_ty))
+        } else {
+            // An index to a function type
+            wasm.read_var_i33()
+                .and_then(|idx| idx.try_into().map_err(|_| Error::InvalidFuncTypeIdx))
+                .map(BlockType::Type)
+        }
```

Comment:
I think there might be an error here, coming from a general issue with the current WasmReader-pattern.
When `ValType::read` is called, it might consume some bytes and only then realize the input is invalid. Thus it returns `Err(_)`, however the previously consumed bytes are never returned to the `WasmReader`.
In this case this code would then go into the `else` block and try to read an `i33` from the now different starting index.

In my implemention of this method I added the comment: `// FIXME: Use transactions for ValType::read`, which is missing here. I suggest something along the lines of:
```suggestion
        if wasm.peek_u8()? as i8 == 0x40 {
            // Empty block type
            let _ = wasm.read_u8().unwrap_validated();
            Ok(BlockType::Empty)
        } else if let Ok(val_ty) = wasm.handle_transaction(|wasm| ValType::read(wasm)) {
            // No parameters and given valtype as the result
            Ok(BlockType::Returns(val_ty))
        } else {
            // An index to a function type
            wasm.read_var_i33()
                .and_then(|idx| idx.try_into().map_err(|_| Error::InvalidFuncTypeIdx))
                .map(BlockType::Type)
        }
```

Reply:
Please do this for both `BlockType::read` and `BlockTyp::read_unvalidated`

Reply:
Please do the previous change for both `BlockType::read` and `BlockTyp::read_unvalidated`